beacon-chain/db/filesystem/data_column.go

```diff
@@ -0,0 +1,1012 @@
+	errWrongVersion                         = errors.New("wrong version")
+	errWrongBytesHeaderRead                 = errors.New("wrong number of bytes header read")
+	errTooManyDataColumns                   = errors.New("too many data columns")
+	errWrongSszEncodedDataColumnSidecarSize = errors.New("wrong SSZ encoded data column sidecar size")
+	errDataColumnSidecarsFromDifferentSlots = errors.New("data column sidecars from different slots")
+)
+
+type (
+	// DataColumnStorage is the concrete implementation of the filesystem backend for saving and retrieving DataColumnSidecars.
+	DataColumnStorage struct {
+		base            string
+		retentionEpochs primitives.Epoch
+		fs              afero.Fs
+		cache           *dataColumnStorageSummaryCache
+		DataColumnFeed  *event.Feed
+		muChans         map[[fieldparams.RootLength]byte]*muChan
+		mu              sync.Mutex // protect muChans
+		pruneMu         sync.RWMutex
+	}
+
+	// DataColumnStorageOption is a functional option for configuring a DataColumnStorage.
+	DataColumnStorageOption func(*DataColumnStorage) error
+
+	// DataColumnIdent is the unique identifier for a data column sidecar.
+	DataColumnsIdent struct {
+		Root    [fieldparams.RootLength]byte
+		Epoch   primitives.Epoch
+		Indices []uint64
+	}
+
+	storageIndices struct {
+		indices [mandatoryNumberOfColumns]byte
+		count   int64
+	}
+
+	metadata struct {
+		indices                         *storageIndices
+		sszEncodedDataColumnSidecarSize uint32
+		fileSize                        int64
+	}
+
+	fileMetadata struct {
+		period    uint64
+		epoch     primitives.Epoch
+		blockRoot [fieldparams.RootLength]byte
+	}
+
+	muChan struct {
+		mu      *sync.RWMutex
+		toStore chan []blocks.VerifiedRODataColumn
+	}
+)
+
+// WithDataColumnBasePath is a required option that sets the base path of data column storage.
+func WithDataColumnBasePath(base string) DataColumnStorageOption {
+	return func(b *DataColumnStorage) error {
+		b.base = base
+		return nil
+	}
+}
+
+// WithDataColumnRetentionEpochs is an option that changes the number of epochs data columns will be persisted.
+func WithDataColumnRetentionEpochs(e primitives.Epoch) DataColumnStorageOption {
+	return func(b *DataColumnStorage) error {
+		b.retentionEpochs = e
+		return nil
+	}
+}
+
+// WithDataColumnFs allows the afero.Fs implementation to be customized.
+// Used by tests to substitute an in-memory filesystem.
+func WithDataColumnFs(fs afero.Fs) DataColumnStorageOption {
+	return func(b *DataColumnStorage) error {
+		b.fs = fs
+		return nil
+	}
+}
+
+// NewDataColumnStorage creates a new instance of the DataColumnStorage object. Note that the implementation of DataColumnStorage may
+// attempt to hold a file lock to guarantee exclusive control of the data column storage directory, so this should only be
+// initialized once per beacon node.
+func NewDataColumnStorage(ctx context.Context, opts ...DataColumnStorageOption) (*DataColumnStorage, error) {
```

Comment:
Can this PR include the `node.go` code that wires up this storage? IMO it is important to see the initialization in context. For instance, I can't tell where WarmCache would run - does it run once at startup and block like the existing blob storage?

Reply:
I had some complains in former PRs that they were already too big.
==> I decided to adopt a bottom up approach, starting by merging all packages, and then using them in eventual PRs.

But yes I agree in this case some context is missing.
However, the context can be easily found in the `peerDAS` branch.

For example, the data column storage is created [here](https://github.com/OffchainLabs/prysm/blob/8d812d5f0e6008a2c756279cc3a41f05dd1f7a7d/beacon-chain/node/node.go#L198-L205) (just after the blob storage), and `WarmCache` is used [here](https://github.com/OffchainLabs/prysm/blob/8d812d5f0e6008a2c756279cc3a41f05dd1f7a7d/beacon-chain/node/node.go#L298-L299) (just after the blob cache warming).

Reply:
Thanks for the references!